### PR TITLE
Refs #34038 -- Improve link readability and identifiability

### PR DESCRIPTION
**Pull request by yushanfans2233**
#### Before
before-light

#### After
after-light

#### Before
before-dark

#### After
after-dark

### CODE REVIEW

**Comment by sarahboyce** [comment]
If you like, you can add a screenshot within the `test_user_change_password` test (something like `self.take_screenshot("user_change")`) which will also add this page with light dark mode etc in our tests 👍

**Comment by yushanfans2233** [comment]
@sarahboyce done, PTAL 😄

**Comment by sarahboyce** [comment]
Thank you @yushanfans2233! This looks right but for some reason I can't see another screenshot generated 🤔 can you run the new selenium test with screenshots locally (https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/unit-tests/#running-the-selenium-tests) and see if it created screenshots?

**Comment by yushanfans2233** [comment]
I ran the command `.\runtests.py --parallel=1 --selenium=chrome --screenshots`.
Then I see the following screenshots created:
- test_user_change_password_dark-user_change.png
- test_user_change_password_desktop_size-user_change.png
- test_user_change_password_mobile_size-user_change.png
- test_user_change_password_rtl-user_change.png

**Review by sarahboyce — APPROVED** [review]
Needs a review from the accessibility team but this looks good to me ⭐

**Comment by yushanfans2233** [comment]
Thanks for your help, really appreciate it 😸 .

File: tests/admin_views/tests.py
Comment (felixxm) [inline tests/admin_views/tests.py:6730]: This test doesn't have any assertions. Instead I'd add `take_screenshot` to one of existing tests, e.g. `UserAdminTest.test_save_button()`.

File: django/contrib/admin/static/admin/css/base.css
Comment (sabderemane) [inline django/contrib/admin/static/admin/css/base.css:106]: @knyghty @thibaudcolas I have one concern about this change, I haven't tested this change yet and can't see the screenshot on my phone, but it got me thinking about titles that can also be links, are they also considered underlined or just the content?

**Review by sabderemane — COMMENTED** [review]
Looks good! Thank you for your contribution @yushanfans2233 😄
I just add a comment for the accessibility team

File: django/contrib/admin/static/admin/css/base.css
Comment (felixxm) [reply django/contrib/admin/static/admin/css/base.css:106]: As far as I'm aware, all links in the admin will be underlined, so this is a big UI change.

Comment (sabderemane) [reply django/contrib/admin/static/admin/css/base.css:106]: Argh, sorry I should have rephrase, I wanted to say : "do we want them also underlined or just the content?"
I don't know if it's already the case according the style all links in the admin will be underlined as you said, but I think we should agree on what we want before to merge this PR

**Review by knyghty — COMMENTED** [review]
Thanks for the work here, it's definitely needed. Also sorry for the delay.

Clearly the accessibility of the links in the body is improved. I think we do need some overrides though. The main one is that buttons that act as links are currently underlined, e.g. the "delete" button and the "Add Foo" button on the list page as well as the view on site and history buttons.

There are some where I'm less certain. For example the title text (default "Django administration" probably doesn't _need_ to be underlined. I'd also argue that the links in the navigation sidebar don't really need to be underlined, either. But again, interested in a second opinion here.

I also see a link that should be underlined, but isn't:

Screenshot 2023-11-02 at 17 41 30

I think unrelated to this PR, but I also notice the log out link in the header has a fainter underline than the rest of the links, because it's using `border-bottom` rather than `text-decoration`:

Screenshot 2023-11-02 at 17 46 06

**Comment by yushanfans2233** [comment]
Here are two alternative solutions:

### Solution 1
Apply an underline and blue color to all `<a>` tags, and use a CSS class to remove this style from specific locations where it is not needed.

- Pros: Ensures that no links requiring improvement are missed, even if new ones are added in the future.
- Cons: May require modifying a significant number of places.

### Solution 2
Introduce a new CSS class called `.a11y-link` and apply it to the links that require improved readability and identifiability.

- Pros: Provides more precise targeting without affecting existing elements.
- Cons: There is a potential risk of missing some existing links, as well as future additions, when attempting to identify and modify them all.

**Comment by knyghty** [comment]
@yushanfans2233 personally I'd give option 1 a go. It's better to accidentally add an underline to something than it is to accidentally forget one, and if there are any they can be fixed over time as they're noticed.

**Comment by yushanfans2233** [comment]
Hello everyone, I have made some necessary fixes based on the following principles:

1. To indicate its functionality in navigating to a new page or an anchor within the current page, the `<a>` element should be styled with color and underlining.

2. When the `<a>` element is utilized as a button or appears similar to a button, it should not be styled with color and underlining.

File: tests/admin_views/tests.py
Comment (yushanfans2233) [reply tests/admin_views/tests.py:6730]: `UserAdminTest` is not a `AdminSeleniumTestCase`, I can't take screenshot in it.

**Review by knyghty — COMMENTED** [review]
One thing I'm not sure about. Maybe it's fine as it is or maybe not.

At the moment the filters don't have underline, and the only blue ones are the active filters. Clearly we need to show which filter is active. I'm just not sure which is best. Maybe it's fine how it is, but it is a bit hard to discern the text there are links.

Screenshot 2023-11-05 at 16 43 58

The Today and Now controls also need _something_ but not sure what. At the moment it's very unclear you can interact with them. But they don't really act like links to other pages, they are more like buttons. Again, not sure the best way forward here:

Screenshot 2023-11-05 at 16 47 40

**Comment by yushanfans2233** [comment]
> One thing I'm not sure about. Maybe it's fine as it is or maybe not.
> 
> At the moment the filters don't have underline, and the only blue ones are the active filters. Clearly we need to show which filter is active. I'm just not sure which is best. Maybe it's fine how it is, but it is a bit hard to discern the text there are links.
> 
> Screenshot 2023-11-05 at 16 43 58
> The Today and Now controls also need _something_ but not sure what. At the moment it's very unclear you can interact with them. But they don't really act like links to other pages, they are more like buttons. Again, not sure the best way forward here:
> 
> Screenshot 2023-11-05 at 16 47 40

Yes, I completely agree with your viewpoint. It is important to enhance the visibility of these elements and make it clear to users that they are clickable functional elements. However, I believe we should create a new ticket to address this improvement as it is not directly related to the changes made in this PR.

**Comment by knyghty** [comment]
For the sidebar filter, I can maybe agree, however there is a regression for the datetime inputs as this PR removes the colour. See this example from 4.2:

Screenshot 2023-11-06 at 14 40 19

**Comment by yushanfans2233** [comment]
Yes, there is a regression on this link, but I don't believe it was introduced by my PR. I am running an admin site with the latest code from the Django main branch, and it does not display as blue.

image

If you want, I can fix it in this PR.

**Comment by Keramblock** [comment]
Any news on this PR? We also need that =(

**Comment by nessita** [comment]
> Any news on this PR? We also need that =(

As per the latest activity, it seems that this PR needs a review from the @django/accessibility team.

**Review by knyghty — CHANGES_REQUESTED** [review]
We had a good chat about this at our latest @django/accessibility meeting.

We agreed that for now, we should just fix the colours and not add the underline.

We should still do the underline part, but as a new ticket, because:
- The changes can easily be done separately.
- Because of the difficulty in deciding what should and shouldn't be underlined, it's better to get the less contentious change out first.

**Comment by nessita** [comment]
> We had a good chat about this at our latest @django/accessibility meeting.
> 
> We agreed that for now, we should just fix the colours and not add the underline.
> 
> We should still do the underline part, but as a new ticket, because:
> 
>     * The changes can easily be done separately.
> 
>     * Because of the difficulty in deciding what should and shouldn't be underlined, it's better to get the less contentious change out first.

Thank you @knyghty. I agree that separating the link color change from the underlining work is a good path forward. Therefore I recommend:

* File a new ticket for the color link regression, so I can triage it as a release blocker and we can provide the minimal fix suitable for backporting, and
* Potentially re-title the original [ticket #34038](https://code.djangoproject.com/ticket/34038) to focus on the underlines.

What do you think?

**Comment by knyghty** [comment]
@nessita sounds good on my end

**Comment by yushanfans2233** [comment]
Dear Django community,

I regret to inform you that due to my current busy schedule, I am unable to continue working on this PR for the time being. However, please know that I remain eager to assist if needed on any other project within the community.
So I suggest to close this PR and transfer the ticket to others.

Thank you for your understanding.

**Comment by nessita** [comment]
> Dear Django community,
> 
> I regret to inform you that due to my current busy schedule, I am unable to continue working on this PR for the time being. However, please know that I remain eager to assist if needed on any other project within the community. So I suggest to close this PR and transfer the ticket to others.
> 
> Thank you for your understanding.

No worries @yushanfans2233, we appreciate everything you have done so far! Thank you for letting us know.

**Pull request closed without merging**
